### PR TITLE
Added the EXISTS CIP

### PR DESCRIPTION
**Pull request by systay**
Link to a viewable CIP: https://github.com/systay/cypher/blob/69725d2e61b73729705cab483e3012c2df845861/cip/CIP2015-05-13-EXISTS.adoc

### CODE REVIEW

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Authors:* Andrés Taylor <andres@neotechnology.com>
+
+
+[abstract]
+.Abstract
+--
+This CIP introduces EXISTS, a general function that can test for existence of properties, simple patterns and full subqueries.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "for existence" -> "for the existence"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+== Background
+
+In a way, Cypher already has existential subqueries - pattern predicates allow checking for the existence of subgraphs. This form is helpful, but not exhaustive in the forms of subqueries that users want to be able to express.
+
+== Proposal
+
+To make this feature more powerful, add a new EXISTS-function, that would exist in two forms:
+ . Property existence checking
+ . Subqueries existence checking
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: Inconsistent pluralization... I suggest "Subquery existence checking"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                ;
+
+subquery exists = "EXISTS", "{", read only clause, { read only clause }, "}" ;
+
+simple subquery exists = "EXISTS", "{", simple match, "}" ;
+
+simple match = pattern, { ",", pattern }, [ "WHERE", predicate ] ;
+
+----
+Note: read only clause above includes WITH but not RETURN.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "read only clause" -> "the read-only clause"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+=== Semantics
+
+All forms of +EXIST(...)+ accomplish the same task - check if a particular pattern exists in the database.
+They are expressions, and as such must be side-effect free. That is, the subqueries in +EXIST+ must not be updating queries.
+
+==== Property exists
+
+In the property checking form, either an identifier or a string expression is used to represent the property key name. If the node/relationship returned has a property with the given name, the +EXISTS(...)+ expression should return true, otherwise it should return false. If the node/relationship expression returns a null value, the +EXISTS(...)+ expression should return null as well.
+
+This form of +EXISTS+ replaces the already existing +has(...)+ function.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "already existing" -> "pre-existing"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+==== Subquery exists
+Identifiers and parameters from the outer query are available in the subquery. Identifiers introduced in the subquery are not available on the outside.
+
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: I guess we leave the term "identifiers" for now... even though we are looking to replace this with "variable"?

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+MATCH (actor:Actor)
+WHERE EXISTS {
+	MATCH (other:Actor)-[:ACTED_IN]->(movie)<-[:ACTED_IN]-(actor)
+	WHERE other.name = actor.name
+}
+RETURN actor
+----
+
+=== Interaction with existing features
+This replaces the already existing +has(...)+ function.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "already existing" -> "pre-existing"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+== What others do
+
+This is very close to what SQL does with its EXISTS functionality.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "close" -> "similar"

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+----
+MATCH (actor)
+WHERE EXISTS(actor["slogan"])
+RETURN actor
+----
+
+
+_Example 2A:_
+
+Find all actors that have acted together with another actor with the same name.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "actors that" -> "actors who" (replace thoughout)

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+MATCH (person:Person)
+WHERE EXISTS {
+  (person)-[:HAS_DOG]->(dog:Dog)
+  WHERE person.name = dog.name
+}
+RETURN person
+
+----
+This proposal also allows for powerful subqueries, for example using aggregation inside the EXISTS query. E.g. find all teams that have at least two members that have worked on projects considered successes.
+
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: "least two members that have worked on projects considered successes." -> "least two members who have worked on successful projects."

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+Identifiers and parameters from the outer query are available in the subquery. Identifiers introduced in the subquery are not available on the outside.
+
+For each matching subgraph evaluated with +EXISTS(...)+, the result value must be true if the subquery finds at least one matching row. If no matches are found, false should be returned.
+
+Regarding nomenclature, +EXISTS(...)+ is a scalar expression - it returns a single boolean value.
+
+==== Simple subquery exists
+When the subquery can be described with a single +MATCH-WHERE+ clause, the +MATCH+ keyword can be omitted.
+
+=== Examples
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: Ensure that the examples have [source, cypher], as per the CREATE CIP

**Comment by systay** [comment]
re-opening this CIP as #10 (comments addressed there)

**Pull request closed without merging**